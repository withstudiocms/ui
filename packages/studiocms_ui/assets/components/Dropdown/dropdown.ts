class DropdownHelper {
	private container: HTMLDivElement;
	private toggleEl: HTMLDivElement;
	private dropdown: HTMLUListElement;

	private alignment: 'start' | 'center' | 'end';
	private triggerOn: 'left' | 'right' | 'both';
	private fullWidth = false;
	private focusIndex = -1;

	active = false;

	/**
	 * A helper function to interact with dropdowns.
	 * @param id The ID of the dropdown.
	 * @param fullWidth Whether the dropdown should be full width. Not needed normally.
	 */
	constructor(id: string, fullWidth?: boolean) {
		this.container = document.getElementById(`${id}-container`) as HTMLDivElement;

		if (!this.container) {
			throw new Error(`Unable to find dropdown with ID ${id}.`);
		}

		this.alignment = this.container.dataset.align as 'start' | 'center' | 'end';
		this.triggerOn = this.container.dataset.trigger as 'left' | 'right' | 'both';

		this.toggleEl = document.getElementById(`${id}-toggle-btn`) as HTMLDivElement;
		this.dropdown = document.getElementById(`${id}-dropdown`) as HTMLUListElement;

		if (fullWidth) this.fullWidth = true;

		this.hideOnClickOutside(this.container);

		this.initialBehaviorRegistration();
		this.initialOptClickRegistration();
	}

	/**
	 * Registers a click callback for the dropdown options. Whenever one of the options
	 * is clicked, the callback will be called with the value of the option.
	 * @param func The callback function.
	 */
	public registerClickCallback = (func: (value: string) => void) => {
		const dropdownOpts = this.dropdown.querySelectorAll('li');

		for (const opt of dropdownOpts) {
			opt.removeEventListener('click', this.hide);

			opt.addEventListener('click', () => {
				func(opt.dataset.value || '');
				this.hide();
			});
		}
	};

	/**
	 * Sets up all listeners for the dropdown.
	 */
	private initialBehaviorRegistration = () => {
		window.addEventListener('scroll', this.hide);
		document.addEventListener('keydown', (e) => {
			if (e.key === 'Escape') this.hide();
		});
		document.addEventListener('astro:before-preparation', () => {
			this.dropdown.classList.remove('initialized');
		});

		if (this.triggerOn === 'left') {
			this.toggleEl.addEventListener('click', this.toggle);
		} else if (this.triggerOn === 'both') {
			this.toggleEl.addEventListener('click', this.toggle);
			this.toggleEl.addEventListener('contextmenu', (e) => {
				e.preventDefault();
				this.toggle();
			});
		} else {
			this.toggleEl.addEventListener('contextmenu', (e) => {
				e.preventDefault();
				this.toggle();
			});
		}

		this.toggleEl.addEventListener('keydown', (e) => {
			if (!this.active) return;

			if (e.key === 'Enter') {
				e.preventDefault();

				const focused = this.dropdown.querySelector('li.focused') as HTMLLIElement;

				if (!focused) {
					this.hide();
					return;
				}

				focused.click();
			}

			if (e.key === 'ArrowDown') {
				e.preventDefault();

				this.focusIndex =
					this.focusIndex === this.dropdown.children.length - 1 ? 0 : this.focusIndex + 1;
			}

			if (e.key === 'ArrowUp') {
				e.preventDefault();

				this.focusIndex =
					this.focusIndex === 0 ? this.dropdown.children.length - 1 : this.focusIndex - 1;
			}

			if (e.key === 'ArrowDown' || e.key === 'ArrowUp') {
				if (this.focusIndex > this.dropdown.children.length - 1) {
					this.focusIndex = 0;
				}

				this.dropdown.querySelector('li.focused')?.classList.remove('focused');

				const newFocus = this.dropdown.children[this.focusIndex] as HTMLLIElement;

				if (!newFocus) return;

				newFocus.classList.add('focused');
				newFocus.focus();
			}
		});
	};

	/**
	 * Registers callbacks to hide the dropdown when an option is clicked.
	 */
	private initialOptClickRegistration = () => {
		const dropdownOpts = this.dropdown.querySelectorAll('li');

		for (const opt of dropdownOpts) {
			opt.addEventListener('click', this.hide);
		}
	};

	/**
	 * A function to toggle the dropdown.
	 */
	public toggle = () => {
		if (this.active) {
			this.hide();
			return;
		}

		this.show();
	};

	/**
	 * A function to hide the dropdown.
	 */
	public hide = () => {
		this.dropdown.classList.remove('active');
		this.active = false;
		this.focusIndex = -1;

		this.dropdown.querySelector('li.focused')?.classList.remove('focused');

		setTimeout(() => this.dropdown.classList.remove('above', 'below'), 200);
	};

	/**
	 * A function to show the dropdown.
	 */
	public show = () => {
		const isMobile = window.matchMedia('screen and (max-width: 840px)').matches;

		const {
			bottom,
			left,
			right,
			width: parentWidth,
			x,
			y,
			height,
		} = this.toggleEl.getBoundingClientRect();
		const { width: dropdownWidth } = this.dropdown.getBoundingClientRect();

		const optionHeight = 43.28;
		const totalBorderSize = 2;
		const margin = 4;

		const dropdownHeight = this.dropdown.children.length * optionHeight + totalBorderSize + margin;

		const CustomRect = {
			top: bottom + margin,
			left,
			right,
			bottom: bottom + margin + dropdownHeight,
			width: isMobile || this.fullWidth ? parentWidth : dropdownWidth, // Account for scaling of animation
			height: dropdownHeight,
			x,
			y: y + height + margin,
		};

		this.active = true;

		if (isMobile || this.fullWidth) {
			this.dropdown.style.maxWidth = `${parentWidth}px`;
			this.dropdown.style.minWidth = 'unset';
			this.dropdown.style.width = `${parentWidth}px`;
			this.dropdown.style.left = `calc(${parentWidth / 2}px - ${CustomRect.width / 2}px)`;
		} else {
			if (this.alignment === 'end') {
				this.dropdown.style.left = `calc(${parentWidth}px - ${CustomRect.width}px)`;
			}

			if (this.alignment === 'center') {
				this.dropdown.style.left = `calc(${parentWidth / 2}px - ${CustomRect.width / 2}px)`;
			}
		}

		if (!this.dropdown.classList.contains('initialized')) {
			this.dropdown.classList.add('initialized');
		}

		if (
			CustomRect.top >= 0 &&
			CustomRect.left >= 0 &&
			CustomRect.bottom <= (window.innerHeight || document.documentElement.clientHeight) &&
			CustomRect.right <= (window.innerWidth || document.documentElement.clientWidth)
		) {
			this.dropdown.classList.add('active', 'below');
			this.focusIndex = -1;
		} else {
			this.dropdown.classList.add('active', 'above');
			this.focusIndex = this.dropdown.children.length;
		}
	};

	/**
	 * A jQuery-like function to hide the dropdown when clicking outside of it.
	 * @param element The element to hide when clicking outside of it.
	 */
	private hideOnClickOutside = (element: HTMLElement) => {
		const outsideClickListener = (event: MouseEvent) => {
			if (!event.target) return;

			if (!element.contains(event.target as Node) && isVisible(element) && this.active === true) {
				// or use: event.target.closest(selector) === null
				this.hide();
			}
		};

		document.addEventListener('click', outsideClickListener);
	};
}

export { DropdownHelper };

// source (2018-03-11): https://github.com/jquery/jquery/blob/master/src/css/hiddenVisibleSelectors.js
const isVisible = (elem: HTMLElement) =>
	!!elem && !!(elem.offsetWidth || elem.offsetHeight || elem.getClientRects().length);
