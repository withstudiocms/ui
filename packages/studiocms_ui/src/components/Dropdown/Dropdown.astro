---
import type { StudioCMSColorway } from '../../utils/colors.js';
import Icon from '../Icon/Icon.astro';
import type { HeroIconName } from '../Icon/iconType.js';
import './dropdown.css';

/**
 * An option in the dropdown.
 */
interface Option {
	/**
	 * The label of the option.
	 */
	label: string;
	/**
	 * The value of the option, returned by the helper when listened for.
	 */
	value: string;
	/**
	 * Whether the option is disabled.
	 */
	disabled?: boolean;
	/**
	 * The color of the option.
	 */
	color?: StudioCMSColorway;
	/**
	 * The icon to display next to the option.
	 */
	icon?: HeroIconName;
	/**
	 * The href to link to when the option is clicked. When given, the option will be rendered as an anchor tag.
	 */
	href?: string;
}

/**
 * The props for the Dropdown component.
 */
interface Props {
	/**
	 * The options to display in the dropdown.
	 */
	options: Option[];
	/**
	 * Whether the dropdown is disabled.
	 */
	disabled?: boolean;
	/**
	 * The ID of the dropdown. Required because of the helper.
	 */
	id: string;
	/**
	 * The alignment of the dropdown, defaults to `center`. Will not work on mobile due to size constraints.
	 */
	align?: 'start' | 'center' | 'end';
	/**
	 * The type of click with which the dropdown is clicked. Defaults to `left`.
	 */
	triggerOn?: 'left' | 'right' | 'both';
	/**
	 * The offset of the dropdown from the trigger element in pixels.
	 */
	offset?: number;
}

const {
	options,
	disabled = false,
	align = 'center',
	id,
	triggerOn = 'left',
	offset = 0,
} = Astro.props;
---
<div 
  class="sui-dropdown-container"
  class:list={[disabled && 'disabled']}
  data-align={align}
  id={`${id}-container`}
  data-trigger={triggerOn}
  transition:persist
  transition:persist-props
>
  <div class="sui-dropdown-toggle" id={`${id}-toggle-btn`}>
    <slot />
  </div>
  <ul 
    class="sui-dropdown" 
    class:list={[align]}
    role="listbox" id={`${id}-dropdown`} 
    transition:persist
    transition:persist-props
    aria-labelledby={`${id}-toggle-btn`}
    style={`--offset: ${offset}px;`}
  >
    {options.map(({ value, disabled, color, label, icon, href }) => (
      <li
        class="sui-dropdown-option"
        data-value={value}
        class:list={[disabled && "disabled", icon && "has-icon", color, href && "has-href"]}
        role="option"
        aria-selected="false"
      >
				{href ? (
          <a href={href} class="sui-dropdown-link sui-dropdown-line-container">
          	{icon && (<Icon width={24} height={24} name={icon} />)}
						<span>{label}</span>
					</a>
				) : (
					<div class="sui-dropdown-line-container">
						{icon && (<Icon width={24} height={24} name={icon} />)}
						<span>{label}</span>
					</div>
				)}
      </li>
    ))}
  </ul>
</div>
