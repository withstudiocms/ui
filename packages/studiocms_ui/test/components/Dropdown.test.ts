import { describe, expect } from 'vitest';
import Dropdown from '../../src/components/Dropdown/Dropdown.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const mockProps = {
	id: 'dropdown',
	options: [
		{ label: 'Default', value: 'default', color: 'default' },
		{ label: 'Primary', value: 'primary', color: 'primary' },
		{ label: 'Success', value: 'success', color: 'success' },
		{ label: 'Warning', value: 'warning', color: 'warning' },
		{ label: 'Danger', value: 'danger', color: 'danger' },
		{ label: 'Info', value: 'info', color: 'info' },
		{ label: 'Monochrome', value: 'mono', color: 'mono' },
		{ label: 'Disabled', value: 'disabled', disabled: true },
		{ label: 'As Link', value: 'custom-href', href: '/' },
	],
};

describe('Dropdown Component', () => {
	test('renders Dropdown component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Dropdown, 'Dropdown', { props: mockProps });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ disabled: true },
		{ align: 'start' },
		{ align: 'center' },
		{ align: 'end' },
		{ triggerOn: 'left' },
		{ triggerOn: 'right' },
		{ triggerOn: 'both' },
		{ offset: 0 },
		{ offset: 10 },
	])('renders Dropdown component with props "%s"', async (props, { renderComponent }) => {
		const result = await renderComponent(Dropdown, 'Dropdown', {
			props: { ...mockProps, ...props },
		});
		expect(result).toMatchSnapshot();
	});
});
