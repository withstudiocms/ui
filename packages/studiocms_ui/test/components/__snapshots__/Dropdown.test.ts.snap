// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dropdown Component > renders Dropdown component correctly 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ align: 'center' }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ align: 'end' }" 1`] = `"<div class="sui-dropdown-container" data-align="end" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown end" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ align: 'start' }" 1`] = `"<div class="sui-dropdown-container" data-align="start" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown start" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ disabled: true }" 1`] = `"<div class="sui-dropdown-container disabled" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ offset: +0 }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ offset: 10 }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 10px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ triggerOn: 'both' }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="both"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ triggerOn: 'left' }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="left"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;

exports[`Dropdown Component > renders Dropdown component with props "{ triggerOn: 'right' }" 1`] = `"<div class="sui-dropdown-container" data-align="center" id="dropdown-container" data-trigger="right"> <div class="sui-dropdown-toggle" id="dropdown-toggle-btn">  </div> <ul class="sui-dropdown center" role="listbox" id="dropdown-dropdown" aria-labelledby="dropdown-toggle-btn" style="--offset: 0px;"> <li data-value="default" class="sui-dropdown-option default" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Default</span> </div> </li><li data-value="primary" class="sui-dropdown-option primary" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Primary</span> </div> </li><li data-value="success" class="sui-dropdown-option success" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Success</span> </div> </li><li data-value="warning" class="sui-dropdown-option warning" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Warning</span> </div> </li><li data-value="danger" class="sui-dropdown-option danger" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Danger</span> </div> </li><li data-value="info" class="sui-dropdown-option info" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Info</span> </div> </li><li data-value="mono" class="sui-dropdown-option mono" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Monochrome</span> </div> </li><li data-value="disabled" class="sui-dropdown-option disabled" role="option" aria-selected="false"> <div class="sui-dropdown-line-container">  <span>Disabled</span> </div> </li><li data-value="custom-href" class="sui-dropdown-option has-href" role="option" aria-selected="false"> <a href="/" class="sui-dropdown-link sui-dropdown-line-container">  <span>As Link</span> </a> </li> </ul> </div>"`;
